docs/design_notes/type_promotion.md

```diff
@@ -0,0 +1,722 @@
+```
+
++++ {"id": "FZ6ana2UooNh"}
+
+However, these advantages comes with a few tradeoffs:
+
+- mixed float/integer promotion is very prone to precision loss: for example, `int64` (with a maximum value of $9.2 \times 10^{18}$) can be promoted to `float16` (with a maximum value of $6.5 \times 10^4$), meaning most representable values will become `inf`.
+- as mentioned above, `f*` can no longer be thought of as a "scalar type", but whether as a different flavor of float64. In JAX's parlance, this is referred to as a *weak type*, in that it is represented as 64-bit, but only weakly holds to this bit width in promotion with a strongly-typed value.
+
+Note that also, this approach still leaves the `uint64` promotion question unanswered, although it is perhaps reasonable to close the lattice by connecting `u64` to `f*`.
+
++++ {"id": "hLAcEHg7Vm-B"}
+
+## Type Promotion in JAX
+
+In designing the type promotion semantics of JAX, we kept in mind many of these ideas, and leaned heavily on a few things:
+
+1. We chose to constrain JAX's type promotion semantics to graphs that satisfy the lattice property: this is to ensure associativity and commutativity, but also to allow the semantics to be compactly described in a DAG, rather than requiring a large table.
+
+2. We leaned toward semantics that avoid inadvertent promotion to higher-precision, particularly when it comes to float values, in order to benefit computation on accelerators.
+
+3. We were fine accepting potential loss of precision (but not loss of magnitude) in mixed type promotion if it were required to maintain (1) and (2)
+
+With this in mind, JAX has adopted Option 3. Or rather, a slightly modified version of Option 3 that draws the connection between `u64` and `f*`, in order to create a true lattice.
+Rearranging the nodes for clarity, JAX's type promotion lattice then looks like this:
+
+```{code-cell}
+:cellView: form
+:id: I5_GcCGwXMDV
+:tags: [hide-input]
+
+#@title
+import networkx as nx
+import matplotlib.pyplot as plt
+lattice = {
+  'i*': ['u8', 'i8'], 'f*': ['c*', 'f16', 'bf16'], 'c*': ['c64'],
+  'u8': ['u16', 'i16'], 'u16': ['u32', 'i32'], 'u32': ['u64', 'i64'], 'u64': ['f*'],
+  'i8': ['i16'], 'i16': ['i32'], 'i32': ['i64'], 'i64': ['f*'],
+  'f16': ['f32'], 'bf16': ['f32'], 'f32': ['f64', 'c64'], 'f64': ['c128'],
+  'c64': ['c128']
+}
+graph = nx.from_dict_of_lists(lattice, create_using=nx.DiGraph)
+pos = {
+  'i*': [-1.25, 0.5], 'f*': [4.5, 0.5], 'c*': [5, 1.5],
+  'u8': [0.5, 0], 'u16': [1.5, 0], 'u32': [2.5, 0], 'u64': [3.5, 0],
+  'i8': [0, 1], 'i16': [1, 1], 'i32': [2, 1], 'i64': [3, 1],
+  'f16': [5.75, 0.8], 'bf16': [5.75, 0.2], 'f32': [7, 0.5], 'f64': [8, 0.5],
+  'c64': [7.5, 1.5], 'c128': [8.5, 1.5],
+}
+fig, ax = plt.subplots(figsize=(10, 4))
+ax.set_ylim(-0.5, 2)
+nx.draw(graph, with_labels=True, node_size=1500, node_color='lightgray', pos=pos, ax=ax)
+# ax.patches[12].set_linestyle((0, (2, 4)))
+```
+
```

Comment:
Updating removing the u64->f* edge?

Reply:
Yes, this was intentional. I discuss u64 more below.